clippy_lints/src/formatting.rs

```diff
@@ -245,12 +245,10 @@ fn is_block(expr: &ast::Expr) -> bool {
     }
 }
 
-/// Match `if` or `if let` expressions and return the `then` and `else` block.
```

Comment:
The comment should stay the same

Reply:
we probably should just stop using this function and match, really

Reply:
Why? `if let` expressions are not a separate syntactic construct anymore.

Reply:
They are from the point of view of Rustaceans, at least for now.

The implementation is not the same thing as the pedagogical model; which is still the same for now.

Reply:
But this is the implementation (in Clippy), not the pedagogical model (in Clippy docs).

(As for the pedagogical mode, it needs to change with the language.)

Reply:
What I'm saying is that this would not be clear to clippy contributors.

This is in the implementation of _clippy_, not the compiler.

> (As for the pedagogical model, it needs to change with the language.)

But the language hasn't gone there yet. The feature is logically the chaining of if and if let, and that doesn't necessarily mean that we've gotten rid of the distinction. `let` isn't a full fledged expression in this model, there's still a level of distinction.

The bottom line here is that the function exists for a very particular purpose and changing the comment this way makes this purpose unclear.

Please either get rid of the function (it's not really necessary anymore) or fix the comment.

Reply:
Until we're in a situation where pedagogically we teach if and if let as the same kind of thing, _we are not there yet_, and comments in code should reflect the current situation. Not some future ideal. The point of comments is to help those working on the code.

Reply:
> Please either get rid of the function (it's not really necessary anymore) or fix the comment.

I've changed back the comment since it is your code-base.

That said, I'm not convinced by what you are saying. I very well understand the point of comments, but I think the change in the comment reflects reality, not some future ideal. Namely, if we keep in mind that AST and HIR are two separate languages (that is, "Rust" is a tower of languages, not a single one), including in terms of semver, then `let` is a truly fully fledged expression in the model of AST.

Moreover, clippy contributors touching this code work on the AST of the compiler, not some other structure, and in AST, `if let` does not exist.

Anyways... let's agree to disagree here.

Reply:
> Moreover, clippy contributors touching this code work on the AST of the compiler, not some other structure

Sure, but the AST/HIR frequently doesn't match the way we want to look at things, that's _why_ we have desugar/unsugar functions in the first place! We talk about things in terms of the actual Rust syntax (i.e. the pedagogical model), not the AST/HIR -- we frequently shim over cases where the AST/HIR doesn't match the pedagogical model so that the code makes sense to someone unfamiliar with the precise corners of the Rust AST.

Clippy uses the AST model because it has to, not because it wants to. We've got efforts to move _away_ from the AST/HIR models, even: https://github.com/rust-lang/rust-clippy/pull/3875/files

Reply:
> We talk about things in terms of the actual Rust syntax (i.e. the pedagogical model),

But that's *my* point. 😂 

The *actual* Rust *concrete syntax* supports `let` in arbitrary expression positions on nightly. It's not just a matter of AST representation but what `parser.rs` can produce. Now if you would say that we want to talk about stable Rust, I'd say that would be a good point.

In the specific context of `unsugar_if`, I believe it is also entirely irrelevant whether the condition is a `let` or something else.

Reply:
> The actual Rust concrete syntax supports let in arbitrary expression positions on nightly

No, by syntax I mean the concepts we use to teach and talk about the syntax, not the actual grammar.

Besides, let in arbitrary positions can parse but is forbidden, it's an implementation detail that the parser accepts it. The RFCd feature only allows let expressions within ifs, and the final pedagogical model for that may not be to consider let to be an expression.

> In the specific context of unsugar_if, I believe it is also entirely irrelevant whether the condition is a let or something else.

Sure, but the reason that function exists is to distinguish itself from checking for _just_ normal `if`s.

(Which is why I said we should remove it)